src/tokenserver/db/models.rs

```diff
@@ -61,152 +70,294 @@ impl TokenserverDb {
-            .bind::<Integer, _>(node.downed)
-            .bind::<Integer, _>(node.backoff)
-            .execute(&self.inner.conn)?;
+        Ok(user)
+    }
 
-        let query = r#"
-            SELECT id
-              FROM nodes
+    /// Mark users matching the given email and service ID as replaced.
+    fn replace_users_sync(&self, params: params::ReplaceUsers) -> DbResult<results::ReplaceUsers> {
+        const QUERY: &str = r#"
+            UPDATE users
+               SET replaced_at = ?
              WHERE service = ?
-               AND node = ?
-               AND available = ?
-               AND current_load = ?
-               AND capacity = ?
-               AND downed = ?
-               AND backoff = ?
+               AND email = ?
+               AND replaced_at IS NULL
+               AND created_at < ?
         "#;
-        diesel::sql_query(query)
-            .bind::<Integer, _>(node.service_id)
-            .bind::<Text, _>(&node.node)
-            .bind::<Integer, _>(node.available)
-            .bind::<Integer, _>(node.current_load)
-            .bind::<Integer, _>(node.capacity)
-            .bind::<Integer, _>(node.downed)
-            .bind::<Integer, _>(node.backoff)
-            .get_result::<results::PostNode>(&self.inner.conn)
+        let timestamp = Self::get_timestamp_in_milliseconds();
+
+        diesel::sql_query(QUERY)
+            .bind::<Bigint, _>(timestamp)
+            .bind::<Integer, _>(&params.service_id)
+            .bind::<Text, _>(&params.email)
+            .bind::<Bigint, _>(timestamp)
+            .execute(&self.inner.conn)
+            .map(|_| ())
             .map_err(Into::into)
     }
 
-    #[cfg(test)]
-    fn post_service_sync(&self, service: params::PostService) -> DbResult<results::PostService> {
-        let query = r#"
-            INSERT INTO services (service, pattern)
-            VALUES (?, ?)
+    /// Mark the user with the given uid and service ID as being replaced.
+    fn replace_user_sync(&self, params: params::ReplaceUser) -> DbResult<results::ReplaceUser> {
+        const QUERY: &str = r#"
+            UPDATE users
+               SET replaced_at = ?
+             WHERE service = ?
+               AND uid = ?
         "#;
-        diesel::sql_query(query)
-            .bind::<Text, _>(&service.service)
-            .bind::<Text, _>(service.pattern)
-            .execute(&self.inner.conn)?;
 
-        let query = r#"
-            SELECT id
-              FROM services
+        diesel::sql_query(QUERY)
+            .bind::<Bigint, _>(params.replaced_at)
+            .bind::<Integer, _>(params.service_id)
+            .bind::<Bigint, _>(params.uid)
+            .execute(&self.inner.conn)
+            .map(|_| ())
+            .map_err(Into::into)
+    }
+
+    /// Update the user with the given email and service ID with the given `generation` and
+    /// `keys_changed_at`.
+    fn put_user_sync(&self, params: params::PutUser) -> DbResult<results::PutUser> {
```

Comment:
Analogous legacy Tokenserver DB method is [here](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/assignment/sqlnode/sql.py#L397). Note that my approach is different here: I plan to move any `client_state`-checking logic to the handler. That seems more like "business logic" to me, and I'd prefer to keep the database methods slim.

Reply:
FWIW, some distributed databases charge per MB output, so moving filtering logic into the query can make some sense. For the data we're talking about here, it's not cost effective. (There are also some other storage systems that charge on data collected before filtering is done, and then charge on data output as well, but that's not the case here.)